### PR TITLE
feat: Make containerd restart its own patch

### PR DESCRIPTION
**Pull request by dlipovetsky**
We have multiple patches that configure containerd. We need to restart containerd after all such patches, and we should restart it exactly once.

Although we've been configuring http proxy for the containerd systemd service, we haven't been restarting it, and I believe the configuration wasn't being applied. This PR should fix that, too.
 
**NOTE: This depends on #16 and #17, because the tests fail otherwise. Do not merge until both of these have merged.**

### CODE REVIEW

File: pkg/handlers/generic/mutation/containerdrestart/inject.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/patches"
+	"github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/patches/selectors"
+)
+
+type globalMirrorPatchHandler struct{}
```
Comment (supershal) [inline pkg/handlers/generic/mutation/containerdrestart/inject.go:20]:
```suggestion
type containerdRestartPatchHandler struct{}
```

File: pkg/handlers/generic/mutation/containerdrestart/tests/generate_patches.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+					Path:      "/spec/template/spec/kubeadmConfigSpec/preKubeadmCommands",
+					ValueMatcher: gomega.ContainElements(
+						containerdrestart.ContainerdRestartScriptOnRemoteCommand,
+					),
+				},
+			},
+		},
+		capitest.PatchTestDef{
+			Name:        "restart script and command added to worker node kubeadm config template",
+			RequestItem: request.NewKubeadmControlPlaneTemplateRequestItem(""),
```
Comment (supershal) [inline pkg/handlers/generic/mutation/containerdrestart/tests/generate_patches.go:50]:
```suggestion
			RequestItem: request.NewKubeadmConfigTemplateRequestItem(""),
```

File: pkg/handlers/generic/mutation/containerdrestart/tests/generate_patches.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			ExpectedPatchMatchers: []capitest.JSONPatchMatcher{
+				{
+					Operation: "add",
+					Path:      "/spec/template/spec/kubeadmConfigSpec/files",
```
Comment (supershal) [inline pkg/handlers/generic/mutation/containerdrestart/tests/generate_patches.go:54]:
```suggestion
					Path:      "/spec/template/spec/files",
```

File: pkg/handlers/generic/mutation/containerdrestart/tests/generate_patches.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+					ValueMatcher: gomega.ContainElements(
+						gomega.HaveKeyWithValue(
+							"path", containerdrestart.ContainerdRestartScriptOnRemote,
+						),
+					),
+				},
+				{
+					Operation: "add",
+					Path:      "/spec/template/spec/kubeadmConfigSpec/preKubeadmCommands",
```
Comment (supershal) [inline pkg/handlers/generic/mutation/containerdrestart/tests/generate_patches.go:63]:
```suggestion
					Path:      "/spec/template/spec/preKubeadmCommands",
```

File: pkg/handlers/generic/mutation/containerdrestart/inject.go (hunk 1 above)
Comment (supershal) [reply pkg/handlers/generic/mutation/containerdrestart/inject.go]: fix: https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/18/commits/cfe0f05f59a17e913e2530cf2e4f557aaec19bf9

File: pkg/handlers/generic/mutation/containerdrestart/tests/generate_patches.go (hunk 3 above)
Comment (supershal) [reply pkg/handlers/generic/mutation/containerdrestart/tests/generate_patches.go]: fix: https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/18/commits/cfe0f05f59a17e913e2530cf2e4f557aaec19bf9

File: pkg/handlers/generic/mutation/containerdrestart/tests/generate_patches.go (hunk 4 above)
Comment (supershal) [reply pkg/handlers/generic/mutation/containerdrestart/tests/generate_patches.go]: fix: https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/18/commits/cfe0f05f59a17e913e2530cf2e4f557aaec19bf9

**Review by supershal — DISMISSED** [review]

**Comment by supershal** [comment]
fixed unit test and rebased: https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/commit/17ce0a8a9abc9d79ae67336bd592b2bda744859b

File: pkg/handlers/generic/mutation/handlers.go [hunk 5]
```diff
@@ -30,5 +31,10 @@ func MetaMutators(mgr manager.Manager) []mutation.MetaMutator {
 		mirrors.NewPatch(mgr.GetClient()),
 		calico.NewPatch(),
 		users.NewPatch(),
+
+		// Some patches may have changed containerd configuration.
+		// We must restart containerd for the configuration to take effect.
+		// Therefore, we must apply this patch last.
+		containerdrestart.NewPatch(),
```
Comment (dkoshkin) [inline pkg/handlers/generic/mutation/handlers.go:38]: IIUC the restart will be added all the time, not just when some other patch touches Containerd? Thats totally fine if thats the case, but I think we should leave a comment here about it.

Comment (supershal) [reply pkg/handlers/generic/mutation/handlers.go:38]: yes. We did decided to mimic konvoy behavior where we apply restart regardless any patch touched containerd or not.

Comment (supershal) [reply pkg/handlers/generic/mutation/handlers.go:38]: @dkoshkin WDYT
```go
// Containerd restart and readiness altogether could take ~5s.
// We want to keep patch independent of each other and not share any state.
// Therefore, We must always apply this patch regardless any other patch modified containerd configuration.
```

**Review by dkoshkin — APPROVED** [review]

**Review by supershal — APPROVED** [review]

**Comment by supershal** [comment]
rebased from main after https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/32 was merged.

**Pull request merged**